### ISSUE TITLE
Spring Content doesn't work correctly when using a RepositoryRestConfigurer with an Entity Lookup to customise the id part of URI

### ISSUE THREAD

**Issue by davidm76** [paulcwarren/spring-content#76]
Hi @paulcwarren ,

I'm using `RepositoryRestConfigurer` to customise a spring data rest repository to use a natural id while, behind the schemes still having an auto-generated primary key.

Note, an application demoing this issue can be found here:
https://github.com/davidm76/spring-data-rest-issue

Version is 1.0.0.M1 with Spring Boot 2.2.0.M5. 

Say I've got an entity that looks link this:

```java
@Data
@Entity
class User {

    @Id
    @GeneratedValue
    private long id;

    @NaturalId
    private String userName;

    private String fullName;
}
```

I can use `RepositoryRestConfigurer` so that I can use the username as the id in Spring Data Rest:

```java
@Component
class RepositoryCustomization implements RepositoryRestConfigurer {

    @Override
    public void configureRepositoryRestConfiguration(RepositoryRestConfiguration config) {

        config.withEntityLookup().
                forRepository(UserRepository.class, User::getUserName, UserRepository::findByUserName);
    }
}
```

With this in place, I can create a user like this:

```bash
curl -X POST http://localhost:8080/users/ -H 'Content-Type: application/json' -d '{"userName" : "davidm","fullName" : "David MacDonald"}
```

And I get the following response:

```json
{
  "userName" : "davidm",
  "fullName" : "David MacDonald",
  "_links" : {
    "self" : {
      "href" : "http://localhost:8080/users/davidm"
    },
    "user" : {
      "href" : "http://localhost:8080/users/davidm"
    }
  }
}
```
Note that `href` refers to natural id davidm and not the generated id.

I can delete the user like so, using the username:

```
 curl -X DELETE http://localhost:8080/users/davidm

```

Say I want to store an avatar with the user and I introduce a content store.

So I change User to be:
```java
@Data
@Entity
class User {

    @Id
    @GeneratedValue
    private long id;

    @NaturalId
    private String userName;

    private String fullName;

    @ContentId
    private String contentId;

    @ContentLength
    private long contentLength;

    @MimeType
    private String mimeType;

}
```

And I add an avatar content store:

```java
public interface AvatarContentStore extends ContentStore<User, String> {
}
```

Now when I create a user I get the following response:

```
{
  "userName" : "davidm",
  "fullName" : "David MacDonald",
  "contentId" : null,
  "contentLength" : 0,
  "mimeType" : null,
  "_links" : {
    "self" : {
      "href" : "http://localhost:8080/users/davidm"
    },
    "user" : [ {
      "href" : "http://localhost:8080/users/davidm"
    }, {
      "href" : "http://localhost:8080/users/1"
    } ],
    "users" : {
      "href" : "http://localhost:8080/users/1"
    }
  }
}
```

Notice there is an array of hrefs for the user, one with the natural id (username) and one with the internal id (the int).

Now when I try to delete the user I get the following exception:

```
java.lang.NumberFormatException: For input string: "davidm"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65) ~[na:1.8.0_212]
	at java.lang.Long.parseLong(Long.java:589) ~[na:1.8.0_212]
	at java.lang.Long.valueOf(Long.java:803) ~[na:1.8.0_212]
	at org.springframework.util.NumberUtils.parseNumber(NumberUtils.java:214) ~[spring-core-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.core.convert.support.StringToNumberConverterFactory$StringToNumber.convert(StringToNumberConverterFactory.java:62) ~[spring-core-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.core.convert.support.StringToNumberConverterFactory$StringToNumber.convert(StringToNumberConverterFactory.java:49) ~[spring-core-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.core.convert.support.GenericConversionService$ConverterFactoryAdapter.convert(GenericConversionService.java:436) ~[spring-core-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.core.convert.support.ConversionUtils.invokeConverter(ConversionUtils.java:41) ~[spring-core-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.core.convert.support.GenericConversionService.convert(GenericConversionService.java:191) ~[spring-core-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.core.convert.support.GenericConversionService.convert(GenericConversionService.java:174) ~[spring-core-5.2.0.RC1.jar:5.2.0.RC1]
	at internal.org.springframework.content.rest.controllers.AbstractContentPropertyController.findOne(AbstractContentPropertyController.java:160) ~[spring-content-rest-1.0.0.M1.jar:na]
	at internal.org.springframework.content.rest.controllers.ContentEntityRestController.deleteContent(ContentEntityRestController.java:170) ~[spring-content-rest-1.0.0.M1.jar:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_212]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_212]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_212]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_212]
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:190) ~[spring-web-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:138) ~[spring-web-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:105) ~[spring-webmvc-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:887) ~[spring-webmvc-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:792) ~[spring-webmvc-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) ~[spring-webmvc-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1039) ~[spring-webmvc-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:942) ~[spring-webmvc-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1005) ~[spring-webmvc-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.web.servlet.FrameworkServlet.doDelete(FrameworkServlet.java:930) ~[spring-webmvc-5.2.0.RC1.jar:5.2.0.RC1]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:666) ~[tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:882) ~[spring-webmvc-5.2.0.RC1.jar:5.2.0.RC1]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741) ~[tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231) ~[tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53) ~[tomcat-embed-websocket-9.0.22.jar:9.0.22]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99) ~[spring-web-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118) ~[spring-web-5.2.0.RC1.jar:5.2.0.RC1]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:92) ~[spring-web-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118) ~[spring-web-5.2.0.RC1.jar:5.2.0.RC1]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:200) ~[spring-web-5.2.0.RC1.jar:5.2.0.RC1]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118) ~[spring-web-5.2.0.RC1.jar:5.2.0.RC1]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202) ~[tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96) [tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:490) [tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139) [tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92) [tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) [tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343) [tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:408) [tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66) [tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:853) [tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1587) [tomcat-embed-core-9.0.22.jar:9.0.22]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) [tomcat-embed-core-9.0.22.jar:9.0.22]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_212]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_212]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-embed-core-9.0.22.jar:9.0.22]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_212]
```

Thanks.

David.

**Comment by paulcwarren** [comment]
Hi David,

Thanks for the example project.

Those additional links are added by Spring Content and Spring Content doesn't support `@NaturalId` atm.  We can add support for them though but that sounds like an aside atm.

Your current issue is that you cant delete the user.  

With this configuration of entity the URI for the DELETEing content and for DELETEing entities are the same so your request is being handled by the Spring Content REST controller, not the Spring Data REST controller since Spring Content controllers are matched against the incoming request *before* Spring Data REST controllers.  

One workaround is to tighten the DELETE request, you should be able to add an Accept header; i.e:

```
curl -X DELETE -H "Accept: application/hal+json" http://localhost:8080/users/davidm
``` 

and this will screen out the SC REST controllers.  Admittedly, this is a bit dumb/counter-intuitive as the Spring Data REST controller doesn't actually return any JSON though.

Probably the better solution is to configure the Spring Content REST controllers to be on a [different baseURI](https://paulcwarren.github.io/spring-content/refs/release/1.0.0.M1/rest-index.html#_changing_the_base_uri).  That would make the URIs different.  I PRed that change back to you [here](https://github.com/davidm76/spring-data-rest-issue/pull/1).

Either should work though.  

Then, because SC doesn't yet support `@NaturalId`, adding content has to be to the Id-based URI:
`curl -X PUT http://localhost:8080/contentApi/users/1 -F 'file=@/tmp/foo'`

and getting again:

`curl http://localhost:8080/contentApi/users/1`

In terms of supporting @NaturalId; a couple of questions for you:
- Would you expect SC to support `@NaturalId`-based URIs for content?  Assuming "yes".
- Does your client use the alps profile and follow the "_links", or does your client "know" what URIs to call?  

HTH
_Paul

**Comment by paulcwarren** [comment]
Ping.  Did those suggestions work around the problem for you @davidm76.  Or is the lack of support for @NaturalId a show stopper for you?

**Comment by davidm76** [comment]
Thanks for the suggestions.
It's not so much the @NaturalId as the `RepositoryRestConfigurer`,  What I was wanting to do was use the name field for the REST API but still have a generated numeric id (for exporting to a small client deployed database).  So I ended up not using the `RepositoryRestConfigurer`.  Instead, I made the name the primary key `@Id` and  added another column to my Postgres database of type `serial` and annotated it with `@Generated`.  This achieved what I needed, but it's not the best implementation.  `serial` for example, works with Postgres, but not with other dbs.  It would be great to see this supported in the future.

Thanks again for your help.
